### PR TITLE
Fix SimplifyConstantIfBranchExecution Bugs

### PR DESCRIPTION
**Pull request by JLLeitschuh**
There were several cases where the `SimplifyConstantIfBranchExecution`
recipie was incorrect.

This fixes those cases and simplifies removing unneeded blocks into
it's own visitor.

### CODE REVIEW

**Comment by sambsnyd** [comment]
@JLLeitschuh if you run `./gradlew licenseFormat` it'll add license headers to the tops of the files you've added, that'll prevent the CI build failures that are complaining about missing license headers

File: rewrite-java/src/main/java/org/openrewrite/java/cleanup/SimplifyConstantIfBranchExecution.java [hunk 1]
```diff
@@ -46,6 +47,8 @@ public String getDescription() {
     }
 
     private static class SimplifyConstantIfBranchExecutionVisitor extends JavaVisitor<ExecutionContext> {
+        private static final UnnecessaryParenthesesVisitor UNNECESSARY_PARENTHESES_VISITOR =
```
Comment (pway99) [inline rewrite-java/src/main/java/org/openrewrite/java/cleanup/SimplifyConstantIfBranchExecution.java:50]: [TreeVisitor](https://github.com/openrewrite/rewrite/blob/6c391098cb5363bc4bdf3316940013e0dfbd8e28/rewrite-core/src/main/java/org/openrewrite/TreeVisitor.java#L50) is not thread-safe, remove this variable declaration and use a new instance for each operation

File: rewrite-java/src/main/java/org/openrewrite/java/cleanup/SimplifyConstantIfBranchExecution.java [hunk 2]
```diff
@@ -55,67 +58,88 @@ public J visitWhileLoop(J.WhileLoop whileLoop, ExecutionContext executionContext
...
-                })), executionContext, getCursor().getParent());
+            bl = (J.Block) new RemoveUnneededBlock.RemoveUnneededBlockStatementVisitor()
+                .visitNonNull(bl, executionContext, getCursor().getParentOrThrow());
+            EmptyBlockStyle style = ((SourceFile) getCursor().firstEnclosingOrThrow(JavaSourceFile.class))
+                .getStyle(EmptyBlockStyle.class);
+            if (style == null) {
+                style = Checkstyle.emptyBlock();
             }
+            bl = (J.Block) new EmptyBlockVisitor<>(style)
+                .visitNonNull(bl, executionContext, getCursor().getParentOrThrow());
```
Comment (pway99) [inline rewrite-java/src/main/java/org/openrewrite/java/cleanup/SimplifyConstantIfBranchExecution.java:69]: these operations should be guarded to prevent modifications that are unrelated to the `SimplifyConstantIfBranchExecution` concerns

File: rewrite-java/src/main/java/org/openrewrite/java/cleanup/RemoveUnneededBlock.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
+package org.openrewrite.java.cleanup;
+
+import org.openrewrite.ExecutionContext;
+import org.openrewrite.Recipe;
+import org.openrewrite.internal.ListUtils;
+import org.openrewrite.java.JavaIsoVisitor;
+import org.openrewrite.java.JavaVisitor;
+import org.openrewrite.java.tree.J;
+
+public class RemoveUnneededBlock extends Recipe {
```
Comment (pway99) [inline rewrite-java/src/main/java/org/openrewrite/java/cleanup/RemoveUnneededBlock.java:10]: can you add an `@Incubating(since = "7.21.0")` here

File: rewrite-test/src/main/kotlin/org/openrewrite/java/cleanup/SimplifyConstantIfBranchExecutionTest.kt [hunk 4]
```diff
@@ -13,7 +13,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-@file:Suppress("StatementWithEmptyBody")
+@file:Suppress("StatementWithEmptyBody", "PointlessBooleanExpression")
```
Comment (JLLeitschuh) [inline rewrite-test/src/main/kotlin/org/openrewrite/java/cleanup/SimplifyConstantIfBranchExecutionTest.kt:16]: Because, yes, I'm fundamentally writing a bunch of "PointlessBooleanExpression" 😆

File: rewrite-java/src/main/java/org/openrewrite/java/tree/J.java [hunk 5]
```diff
@@ -757,6 +757,20 @@ public Block withStatements(List<JRightPadded<Statement>> statements) {
...
+        public static J.Block empty() {
+            return new J.Block(
+                    Tree.randomId(),
+                    Space.EMPTY,
+                    Markers.EMPTY,
+                    JRightPadded.build(false),
+                    Collections.emptyList(),
+                    Space.EMPTY
+            );
+        }
```
Comment (pway99) [inline rewrite-java/src/main/java/org/openrewrite/java/tree/J.java:773]: Can you move this into the `SimplifyConstantIfBranchExecutionVisitor`?

File: rewrite-java/src/main/java/org/openrewrite/java/cleanup/SimplifyConstantIfBranchExecution.java [hunk 6]
```diff
@@ -47,75 +48,94 @@ public String getDescription() {
...
-                bl = maybeAutoFormat(bl, bl.withStatements(ListUtils.flatMap(bl.getStatements(), stmt -> {
-                    if (stmt == removeIf) {
-                        return addStatements;
-                    }
-                    return stmt;
-                })), executionContext, getCursor().getParent());
+            if (bl != block) {
+                bl = (J.Block) new RemoveUnneededBlock.RemoveUnneededBlockStatementVisitor()
+                    .visitNonNull(bl, executionContext, getCursor().getParentOrThrow());
+                EmptyBlockStyle style = ((SourceFile) getCursor().firstEnclosingOrThrow(JavaSourceFile.class))
```
Comment (pway99) [inline rewrite-java/src/main/java/org/openrewrite/java/cleanup/SimplifyConstantIfBranchExecution.java:57]: `visitBlock` is a fairly hot path, consider calculating the style only once by adding a local variable and getter that will only calculate the style if its null.

**Review by pway99 — CHANGES_REQUESTED** [review]
Thanks @JLLeitschuh!
Just a few requests; avoid calculating the style for every block and move the J.Body.empty() to the visitor.

File: rewrite-java/src/main/java/org/openrewrite/java/tree/J.java (hunk 5 above)
Comment (JLLeitschuh) [reply rewrite-java/src/main/java/org/openrewrite/java/tree/J.java:773]: I can, but why? This is a factory method for creating empty `J.Block` objects. I presume that my visitor will not be the only case where a visitor will want to return an empty `J.Block`.

Comment (pway99) [reply rewrite-java/src/main/java/org/openrewrite/java/tree/J.java:773]: > I can, but why? 

To keep the domain and API as small and straightforward as is possible.  So far, there are more than 700 recipes and less than ten instances where `J.Block` is instantiated with an empty list of statements.

File: rewrite-java/src/main/java/org/openrewrite/java/cleanup/SimplifyConstantIfBranchExecution.java (hunk 6 above)
Comment (pway99) [reply rewrite-java/src/main/java/org/openrewrite/java/cleanup/SimplifyConstantIfBranchExecution.java:57]: in hindsight, the EmtpyBlockStyle is only created when a change is detected so this actually not that expensive

Comment (pway99) [reply rewrite-java/src/main/java/org/openrewrite/java/cleanup/SimplifyConstantIfBranchExecution.java:57]: in hindsight, the EmtpyBlockStyle is only created when a change is detected so this actually not that expensive

**Review by pway99 — APPROVED** [review]
